src/runtime/storage/pouchdb/pouch-db-upsert.ts

```diff
@@ -0,0 +1,86 @@
+/**
+ * @license
+ * Copyright (c) 2017 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import {PouchDB, PouchDbDebug, PouchDbMemory} from '../../../platform/pouchdb-web.js';
+
+export interface UpsertDoc {
+  version: number; // TODO version
+}
+
+export interface UpsertMutatorFn<T extends UpsertDoc> {
+  (arg: T): Promise<T>;
+}
+
+/**
+ * Provides a way to read-modify-write data for a given PouchDB key
+ * based on the following rules:
+ *
+ * - A new entry based on defaultValue is stored if it doesn't exist.
+ * - If the existing entry is available it is fetched.
+ * - A copy of the existing document is pass passed to `mutatorFn`
+ * - If the model is new or mutated by `mutatorFn`, write a new revision.
+ *
+ * @return the current value of data for `docId`
+ * @param db The database where we send our get/put calls.
+ * @param docId The value to get and set
+ * @param mutatorFn A function that can optionally make changes to the existing entry.
+ * @param defaultValue If the docId does not exist use this and also pass through mutatorFn.
+ */
+export async function upsert<T extends UpsertDoc>(
+    db: PouchDB.Database,
+    docId: string,
+    mutatorFn?: UpsertMutatorFn<T>,
+    defaultValue?: PouchDB.Core.NewDocument<T>): Promise<T> {
+
+  // Keep retrying the operation until it succeeds.
+  // TODO(lindner): add backoff and error out if this goes on for too long
+
+  while (1) {
+    let doc: T;
+    let existingDoc = false;
+
+    // Check for an existing document.
+    try {
+      doc = await db.get(docId);
+      existingDoc = true;
+    } catch (err) {
+      if (err.name !== 'not_found') {
+        throw err;
+      }
+      doc = {_rev: undefined, _id: docId, ...defaultValue};
+    }
+
+    // TODO deepClone? Readonly?  For now just use a shallow clone to avoid weirdness..
```

Comment:
What weirdness do you have in mind? To help next reader.

Reply:
I'll document this more.   The mutator receives an object that is modifiable; but after the call there's a check to determine if changes were made.  If the mutator modifies the original object then that will fail.

So we pass in a clone of the object.  Using the spread operator only does a shallow clone though.  In future a deepClone would be preferred to allow the mutator to modify the object any way it seems fit.

Reply:
docs updated.